miri/lib.rs

```diff
@@ -242,25 +251,97 @@ impl<'tcx> Machine<'tcx> for Evaluator<'tcx> {
+        _id: AllocId,
+        _mutability: Mutability,
+    ) -> EvalResult<'tcx, bool> {
+        /*use memory::MemoryKind::*;
         match m {
             // FIXME: This could be allowed, but not for env vars set during miri execution
             Env => err!(Unimplemented("statics can't refer to env vars".to_owned())),
-            _ => Ok(()),
+            _ => Ok(false), // TODO: What does the bool mean?
+        }*/
+        Ok(false)
+    }
+
+    fn init_static<'a>(
+        ecx: &mut EvalContext<'a, 'mir, 'tcx, Self>,
+        cid: GlobalId<'tcx>,
+    ) -> EvalResult<'tcx, AllocId> {
+        // Step 1: If the static has already been evaluated return the cached version
+        if let Some(alloc_id) = ecx.memory.data.mut_statics.get(&cid) {
+            return Ok(*alloc_id);
         }
+
+        let tcx = ecx.tcx.tcx;
+
+        // Step 2: Load mir
+        let mut mir = ecx.load_mir(cid.instance.def)?;
+        if let Some(index) = cid.promoted {
+            mir = &mir.promoted[index];
+        }
+        assert!(mir.arg_count == 0);
+
+        // Step 3: Allocate storage
+        let layout = ecx.layout_of(mir.return_ty().subst(tcx, cid.instance.substs))?;
+        assert!(!layout.is_unsized());
+        let ptr = ecx.memory.allocate(
+            layout.size.bytes(),
+            layout.align,
+            None,
+        )?;
+
+        // Step 4: Cache allocation id for recursive statics
+        assert!(ecx.memory.data.mut_statics.insert(cid, ptr.alloc_id).is_none());
+
+        // Step 5: Push stackframe to evaluate static
+        let cleanup = StackPopCleanup::None;
+        ecx.push_stack_frame(
+            cid.instance,
+            mir.span,
+            mir,
+            Place::from_ptr(ptr, layout.align),
+            cleanup,
+        )?;
+
+        // Step 6: Step until static has been initialized
+        let call_stackframe = ecx.stack().len();
+        while ecx.step()? && ecx.stack().len() >= call_stackframe {
+            if ecx.stack().len() == call_stackframe {
+                let frame = ecx.stack().last().unwrap();
+                let bb = &frame.mir.basic_blocks()[frame.block];
+                if bb.statements.len() == frame.stmt && !bb.is_cleanup {
+                    match bb.terminator().kind {
+                        ::rustc::mir::TerminatorKind::Return => {
+                            for (local, _local_decl) in mir.local_decls.iter_enumerated().skip(1) {
+                                // Don't deallocate locals, because the return value might reference them
+                                // ------------------------------------------------------------
+                                // ||||||||||||| TODO: remove this horrible hack ||||||||||||||
+                                // ------------------------------------------------------------
+                                unsafe { &mut *(frame as *const Frame as *mut Frame) }.storage_dead(local);
```

Comment:
TODO: remove this!

Reply:
Eh, this is an `&` to `&mut` cast, right? This is UB. You might as well deref a dangling pointer here...

Reply:
I know, that is why I want to remove it.